lib/io/pwm.cpp

```diff
@@ -0,0 +1,151 @@
+#include "pwm.hpp"
+
+#include <fcntl.h>   // define O_WONLY and O_RDONLY
+#include <unistd.h>  // close()
+
+namespace hyped::io {
+
+Pwm::Pwm(core::Logger &logger, const PwmOutput pwm_output) : logger_(logger), period_(0)
+{
+  std::string pwm_address = "/sys/class/pwm/" + getPwmFolderName(pwm_output) + "/";
+  // First get the file descriptor for the period file
+  std::string period_address = pwm_address + "period";
+  period_file_               = open(period_address.c_str(), O_WRONLY);
+  if (period_file_ < 0) { logger_.log(core::LogLevel::kFatal, "Failed to open PWM period file"); }
+  // Then get the file descriptor for the duty cycle file
+  std::string duty_cycle_address = pwm_address + "duty_cycle";
+  duty_cycle_file_               = open(duty_cycle_address.c_str(), O_WRONLY);
+  if (duty_cycle_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to open PWM duty cycle file");
+  }
+  // Next get the file descriptor for the polarity file
+  std::string polarity_address = pwm_address + "polarity";
+  polarity_file_               = open(polarity_address.c_str(), O_WRONLY);
+  if (polarity_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to open PWM polarity file");
+  }
+  // Finally get the file descriptor for the enable file
+  std::string enable_address = pwm_address + "enable";
+  enable_file_               = open(enable_address.c_str(), O_WRONLY);
+  if (enable_file_ < 0) { logger_.log(core::LogLevel::kFatal, "Failed to open PWM enable file"); }
+}
```

Comment:
1. Don't do I/O in the constructor, have an initialise function or something.
2. `const`s
3. Return on error (rather than just logging)

One problem I have with NOT doing I/O in the constructor is that it's easy to forget calling the other function. Maybe we want to change it into a static function that returns `std::optional<Pwm>`?

Reply:
won't changing it to a static function mean there is only ever one unique pwm object?

Reply:
i do agree with a separate initalise function tho, its a waste not knowing if a open threw an error during construction. how about if we a `has_initialized_` value that tells us if the initialize function was called or not, i'll commit it in a bit

Reply:
> won't changing it to a static function mean there is only ever one unique pwm object?

not quite; the static function would replace the constructor in the public interface. you could call it as, say, `Pwm::create(logger, pwm_out)`, which would internally construct a `Pwm` object (or fail).

the `static` keyword is a bit overloaded. for class members it simply means that its not part of a particular object but rather of the class. this is similar to Java.

static variables and static functions (outside of classes) are both different to this.

Reply:
ah okay makes sense, will make that change then